sphinx/source/docs/dev_guide/bokehjs.rst

```diff
@@ -121,98 +121,231 @@ Use ``node make help`` to list all available commands.
-6. If tests failed, go to BokehJS's GitHub_Actions_ page. Find the most recent
-   test run for your PR and download the associated ``bokehjs-report`` artifact.
-7. Unzip the artifact archive at the root of the repository.
-8. Assuming devtools server is running in the background, go to ``/integration/report?platform=name``
-   where ``name`` is either ``linux``, ``macos`` or ``windows`` and review the test output
-   for each platform. If there are no unintentional differences, then commit all
-   new or modified ``*.blf`` and ``*.png`` files under ``test/baselines/{linux,macos,windows}``.
-9. Push your changes to GitHub again and verify that tests pass this time.
-
-.. note::
-
-    Make sure to monitor the state of the ``test/baselines`` directory, so that you
-    don't commit unnecessary files. If you do so, subsequent tests will fail. Reset
-    this directory after every failed test run (``git checkout`` and/or ``git clean``).
+BokehJS uses a series of visual baseline comparison tests. These tests help make sure
+that Bokeh's visual output is consistent with the output expected by design. Any
+BokehJS-related pull requests that result in changes to the visual output generated by
+BokehJS should include visual baseline comparison tests.
+
+In the background, BokehJS' testing framework runs a headless browser and takes
+screenshots of the browser's output. The testing framework then compares the visual
+output to each test's dedicated baseline files.
+
+Each test in ``test:integration`` consists of two types of baseline comparisons:
+
+Textual baseline comparison
+  For each test, the testing framework compares the pixel location of certain elements
+  in the visual output to pixel locations in a baseline file. The pixel locations are a
+  set of bounding boxes. They are stored as plain text in each test's ``.blf`` files.
+
+Visual baseline comparison
+  For each test, the testing suite does a pixel-by-pixel comparison of a screenshot
+  and a baseline image. These baseline images are stored as ``.png`` files. In contrast
+  to textual baseline comparisons, visual baseline comparisons are platform-dependent.
+  Even minor differences in font rendering, for example, will make the pixel-by-pixel
+  comparison fail.
+
+The visual baseline comparison tests are located in the
+:bokeh-tree:`bokehjs/test/integration/` folder and sub-folders.
+:ref:`Bokeh's CI <devguide_testing_ci>` runs these tests on Linux, macOS, and Windows
+environments. The baseline files for each environment are located in the
+:bokeh-tree:`bokehjs/test/baselines/` folder.
+
+Follow these steps to write new visual tests or update existing tests:
+
+1. Create or update visual testing scripts:
+    To write a visual test for BokehJS' testing suite, start by importing the
+    ``display()`` and ``fig()`` functions from the testing framework's ``_util`` module:
+
+    .. code-block:: TypeScript
+
+        import {display, fig} from "../_util"
```

Comment:
This import is correct for a test file in a directory like `bokehjs/test/integration/glyphs`, but there are also tests in the parent directory `bokehjs/test/integration` and for these the import is one character different
```
import {display, fig} from "./_util"
```
I'm not advocating putting both of these in as that will be too detailed, but maybe we need to put in something like `for a test file in directory X ...`